Allow mentors to view parts of the applications of teams applying for their project

klappradla opened the pull request:
This adds some more fields to the `ApplicationDraft` model and respectively also the application form (#579 ) and an additional index and show view for mentors to get insight on parts of the fields of applications for their projects (#581).

What's still missing: I need to clean up a few methods in the `Mentor::Application` model, add some proper view code to the show-view and implement a currently pending controller test.

Quick note on why I used SQL and added a new index to a certain path in the hstore: From what I have seen in the Teams app in terms of view code (especially related to ratings), I prefer to have more complex things unit-testalbe as server-code rather then untested view helpers with global scope.
I could have (and also did first) written the SQL statements with AR, but they just end up far less readable with nasty Node-js like brackets and stuff. The way I have it now, it may be a huge statement, but it's super easy to read and understand (if one knows about hstore).
By only pulling out the fields mentors should have access to, we won't have a *"security problem"* of showing them too much information about the applications they should not have access to.

*(sorry for the weird explanation, I'm kinda in a hurry)*

klappradla commented:
It's kind of hard to test / look at this in the browser *(since it relies on certain data being available)*. The way I did it was:
1. import a dump of the production db
1. update all existing applications to the current season
1. impersonate a mentor with a popular project (with multiple 1st and 2nd choice applications)
1. go to `/mentor/applications`

bitboxer commented:
I don't have access to the production data, but I will review the code changes anyway.

bitboxer commented on app/controllers/application_drafts_controller.rb:
```diff
@@ -81,8 +81,12 @@ def application_draft
+      :project1_id, :project2_id,
+      :plan_project1, :plan_project2,
+      :why_selected_project1, :why_selected_project2,
+      :misc_info, :heard_about_it, :voluntary, :voluntary_hours_per_week, :working_together
```
This is missing fields. If this is on purpose, ignore this comment :wink:

klappradla replied:
Missing field that are part of the form? That should probably not be the case 😉 
I'll have a look at the list again, should not miss any fields I think, thanx!

klappradla replied:
So just took a look at it again - seems like it covers all attributes from the form, so missing the rest should be on purpose ✌️ 
*(btw: wifi and "beinfreiheit" on ICE is awesome 🎉  did not know about this)*

bitboxer replied:
Perfect!

bitboxer commented on app/controllers/mentor/base_controller.rb:
```diff
@@ -0,0 +1,9 @@
+  before_action -> { require_role 'mentor' }
+  # before_action :set_breadcrumbs
+
+  # def set_breadcrumbs
```
is this commented out code needed?

klappradla replied:
It's copied over from somewhere else and I was not sure yet if I could use it for the not yet fully implemented show view - but will probably delete it. Thanx for pointing out!

klappradla commented:
So I guess this is ready now. Thanx @bitboxer for the review 🙏 
@carpodaster / @ramonh or @alicetragedy do you also want to have a look at it?

alicetragedy commented:
Also looks good to me 👍

alicetragedy commented:
@klappradla it looks like there's a merge conflict, but I can try to fix it from the command line 😁 I'd be up for merging as soon as possible, since it was reviewed by Bodo already and I trust his judgement. thanks so much! 🎉

bitboxer commented:
@alicetragedy small warning: since I didn't had a production dump at the time of reviewing this, I haven't had the opportunity to try it out, I just reviewed the code :wink: .

alicetragedy commented:
@bitboxer yeah, I saw this :D I also haven't had time to try it out 🙈

klappradla commented:
Sorry for the silence - I just returned from Berlin. I can record a tiny demo of how things look like either tonight or tomorrow if u like @alicetragedy. I did not yet add any navigation, but now that @bitboxer refactored it, it could be worth a look.

Also for merging: if it's too much of a pain, I also do it myself.

alicetragedy commented:
no worries @klappradla. A small demo would be great if you have time; Maria and I would like to do a bit of a test run with the application this week (with other testers) so I think the earlier a lot of these PRs are merged, the better. :)

klappradla commented:
My video is way too big and way too shitty for uploading it here - I put it on Slack 😉 

Quickly what this all does:
* allow mentor to view a list of applications **from the current season** for their projects (they may have multiple)
* don't allow to view other applications - 404
* don't allow to view a teams other project choice - 404
* the values shown are already adapted for this year's data model

ramonh commented:
I just examined this and the video. Everything looks great to me!

klappradla commented:
Ah, now I get what @alicetragedy meant with *"notes"* 🙈  - saw it somewhere else.
I'll also add this feature, but in a separate PR then.

carpodaster commented on app/models/application_draft.rb:
```diff
@@ -9,6 +9,11 @@ class ApplicationDraft < ActiveRecord::Base
   STUDENT0_CHAR_LIMITED_FIELDS = Student::CHARACTER_LIMIT_FIELDS.map { |m| "student0_#{m}" }
   STUDENT1_CHAR_LIMITED_FIELDS = Student::CHARACTER_LIMIT_FIELDS.map { |m| "student1_#{m}" }
 
+  PROJECT1_FIELDS = %i(project1 plan_project1 why_selected_project1)
```
Just a minor thing in terms of keeping the project accessible:  `%i` isn't widely used. Since it just saves a few extra colons, I'd vote for being more explicit and write the literal symbols instead.

klappradla replied:
✔️

carpodaster commented on app/models/creates_application_from_draft.rb:
```diff
@@ -1,4 +1,5 @@
 class CreatesApplicationFromDraft
+  PROJECT_FIELDS = ApplicationDraft::PROJECT_FIELDS.map(&:to_s)
```
It should work with symbols, we could probably spare the #to_s

klappradla replied:
Unfortunately not - it does not have *"indifferent access"* and since it's serialized down into an `hstore` column, it does only work with String.

carpodaster commented on app/models/mentor/application.rb:
```diff
@@ -0,0 +1,118 @@
+      def find(id:, projects:, choice: 1, season: Season.current)
+        params = attrs_for(choice: choice).merge(id: id, project_ids: projects.ids, season_id: season.id)
+        query  = [sql_statement_find, params]
+        data   = DB.select_one(sanitize_sql(query))
```
How about this:
```ruby
data   = DB.select_one(sanitize_sql(query)) || fail ActiveRecord::RecordNotFound
```

klappradla replied:
Oh nice, did not think about this 🙏

carpodaster commented on app/models/mentor/application.rb:
```diff
@@ -0,0 +1,118 @@
+    end
+
+    def student0=(attrs)
+      attrs     = JSON.parse(attrs).instance_eval(&studentize)
```
Holy :cow:  – very elegant!

It took me some time to wrap my mind around it, though, what with the 3 levels of indirection. Do you see a way to rewrite this in a less inside-out object modification manner?

Does this work:

```ruby
def student1=(attrs)
  attrs = JSON.parse(studentize(attrs))
end
# …
def studentize(attrs)
  attrs.keys.each{ |k| attrs[k.gsub(/student(0|1)_application_/, '')] = attrs.delete(k) } }
end
```

But that still modifies the passed `attrs` value, no? It would rather work with `#dup`s. Or is the mutation a desired/required effect?

klappradla replied:
Interesting, I wrote it like this on purpose, cause I thought it would be **better** readable, since you can basically read it from left to right and have every step visible. To me the other approach looked more inside-out, since you have to wrap in in brackets... But if it looked more familiar to you on the first glance, I can easily change it.

And yes, I went for mutation on purpose. The whole process is basically just mapping database columns to model attributes. So there's no point in allocating memory for the intermediate steps and have their results lying around.

carpodaster replied:
Yeah, I agree with the inside-out brackets (if only we had a pipe-operator in Ruby … :wink: )

I'm just unsure if binding the return value of higher order functions/methods to methods as their block parameter which in turn `#tap`s the (implicit) block arg isn't something anyone with less than 4-5 years of Ruby experience immediately shies away from.

I don't know what weighs more so I'll totally shove the responsibility to make a decision back to you :trollface: 

As for the mutation: I don't think memory usage will be an issue for the Teams App anytime soon. Otoh, I can foresee a very nasty debugging session when we try to figure out where certain keys in the passed `params` hash went to. But I don't have strong feelings here, either. You decide! :)

klappradla replied:
Makes sense! I'll come up with something. Just quickly add the "notes" before now.

carpodaster commented on app/models/mentor/application.rb:
```diff
@@ -0,0 +1,118 @@
+module Mentor
+  class Application
```
Would you mind adding a few sprinkles of (YARD) documentation for future versions of ourselves?

klappradla replied:
✔️

klappradla commented:
So @carpodaster I updated the PR according to your feedback and added the comment feature as well. Sorry it took me so long, I had a deadline at work and then a 2 day project out of town.

Quick note on how I implemented the comments:
**Notes** by mentors on applications for their projects are internally represented as `Comment`s in the mentor namespace -> `Mentor::Comment` and use the `Comment` table under the hood. It was not possible / is super difficult to just use the existing `Comment` model with it's polymorphic `Commentable` association, since there are already comments on applications. Including the mentor's personal notes here would cause confusion in other parts of the application plus interfere in the project proposal logic.
Using the exiting `Note` model is also impossible - they are only tied to a user *(and I did not want to change this)*.

So the `Mentor::Comment` is just a `Comment`, not interfering with other comments and each mentor only has one per `Mentor::Application`, using it as a kind of personal notes field that she can also empty again.

klappradla commented:
Follow up to my last comment about **missing parts** (the navigation): I cleaned up the current navigation and basically replaced it with a classic responsive bootstrap one in order to find a place to *"squeeze"* in this new item I called `Mentor Dashboard`.

In order to not have the navigation overflow one line one displays smaller than "huge", I added to dropdown groups:
1. Summer of Code
  Contains phase-dependent everything around participating at the program, such as `Submit your Project`, `Apply now`, `Projects`, `Community` - grouped into logical groups
1. User (displayed via avatar)
  Contains everything concearning the user: `Profile`, `Sign out / Stop inpersonating`, links to dashboards depending on phase and role, etc.
  *(When logged out, there's just a button to log in, no dropdown)*

Tiny demo:
![navigation](https://cloud.githubusercontent.com/assets/3491815/22444555/beecf9a6-e743-11e6-8af4-3d1622c77fc7.gif)

I based these changes on the state of this branch. So one this is merged, I can open up another PR for the navigation - and we can continue the discussion about it there. Fine?

F3PiX commented:
👏  👏   🎬  <- = clapper 
Awesome!!

alicetragedy commented:
This looks so good!!! thank you! 😍

alicetragedy commented:
Shall we merge this? 🎉

carpodaster commented:
Aye, let's do it!

untitaker commented:
Thanks, this came just in time ^^
